### PR TITLE
linux/bpf: regenerate help function with alignment to make compiler happy

### PR DESCRIPTION
**Pull request by tw4452852**

### CODE REVIEW

**Comment by rohlem** [comment]
Does it still make sense then that the default alignment for function pointers on BPF targets is > 1?
Do user-provided functions have a higher alignment restriction than these special helper functions?

**Comment by tw4452852** [comment]
> Does it still make sense then that the default alignment for function pointers on BPF targets is > 1?

I don't think so.

> Do user-provided functions have a higher alignment restriction than these special helper functions?

Yes, I think for these ones that could be `align(sizeof(usize))`

**Comment by alexrp** [comment]
> Does it still make sense then that the default alignment for function pointers on BPF targets is > 1?

We match GCC and LLVM here. If we're going to change this, especially when we use LLVM as our code generator for this target, I would like to see some evidence that changing it won't cause subtle linker bugs.

**Comment by tw4452852** [comment]
I did an experiment with `clang` as follows:

```
> cat t.c

static long (* const bpf_trace_printk)(const char *fmt, int fmt_size, ...) = (void *) 6;

int foo(void) {
        bpf_trace_printk("bar", 3);
        return 0;
}

> clang -S --target=bpfel-freestanding -emit-llvm t.c
> cat t.ll

; ModuleID = 't.c'
source_filename = "t.c"
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "bpfel-freestanding"

@.str = private unnamed_addr constant [4 x i8] c"bar\00", align 1

; Function Attrs: noinline nounwind optnone
define dso_local i32 @foo() #0 {
  %1 = call i64 (ptr, i32, ...) inttoptr (i64 6 to ptr)(ptr noundef @.str, i32 noundef 3)
  ret i32 0
}

attributes #0 = { noinline nounwind optnone "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"frame-pointer", i32 2}
!2 = !{!"clang version 17.0.6"}
```

As you can see, it use `inttoptr` and accordin to

**Comment by tw4452852** [comment]
> %1 = call i64 (ptr, i32, ...) inttoptr (i64 6 to ptr)(ptr noundef @.str, i32 noundef 3)

LLVM use `inttoptr` to case, I think it doesn't assume any alignment requirement.

**Comment by alexrp** [comment]
Well, sure; C and LLVM IR don't represent pointer alignment in the type system like Zig does. So that in itself isn't surprising.

But changing the default function alignment often has broader ABI ramifications. In some cases, the linker may depend on a certain minimum function alignment in order to produce correct binaries. (I don't know if that's the case for BPF.) Here's the relevant code in GCC and LLVM:

* https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/bpf/bpf.h;h=2298e5b05afd94e481d310b3463a020b0c161c62;hb=HEAD#l63
* https://github.com/llvm/llvm-project/blob/dd98ae358b187be32a2e255eba5f91568524b86a/llvm/lib/Target/BPF/BPFISelLowering.cpp#L155-L157

File: lib/std/os/linux/bpf/helpers.zig [hunk 1]
```diff
@@ -11,215 +11,215 @@ const SkFullSock = @compileError("TODO missing os bits: SkFullSock");
...
-pub const skb_store_bytes = @as(*const fn (skb: *kern.SkBuff, offset: u32, from: ?*const anyopaque, len: u32, flags: u64) c_long, @ptrFromInt(9));
-pub const l3_csum_replace = @as(*const fn (skb: *kern.SkBuff, offset: u32, from: u64, to: u64, size: u64) c_long, @ptrFromInt(10));
-pub const l4_csum_replace = @as(*const fn (skb: *kern.SkBuff, offset: u32, from: u64, to: u64, flags: u64) c_long, @ptrFromInt(11));
-pub const tail_call = @as(*const fn (ctx: ?*anyopaque, prog_array_map: *const kern.MapDef, index: u32) c_long, @ptrFromInt(12));
-pub const clone_redirect = @as(*const fn (skb: *kern.SkBuff, ifindex: u32, flags: u64) c_long, @ptrFromInt(13));
-pub const get_current_pid_tgid = @as(*const fn () u64, @ptrFromInt(14));
-pub const get_current_uid_gid = @as(*const fn () u64, @ptrFromInt(15));
-pub const get_current_comm = @as(*const fn (buf: ?*anyopaque, size_of_buf: u32) c_long, @ptrFromInt(16));
-pub const get_cgroup_classid = @as(*const fn (skb: *kern.SkBuff) u32, @ptrFromInt(17));
+pub const map_lookup_elem = @as(*align(1) const fn (map: *const kern.MapDef, key: ?*const anyopaque) ?*anyopaque, @ptrFromInt(1));
```
Comment (InKryption) [inline lib/std/os/linux/bpf/helpers.zig:14]:
```suggestion
pub const map_lookup_elem: *align(1) const fn (map: *const kern.MapDef, key: ?*const anyopaque) ?*anyopaque = @ptrFromInt(1);
```

Comment (tw4452852) [reply lib/std/os/linux/bpf/helpers.zig]: Thanks @InKryption . Applied in the latest version.

**Comment by tw4452852** [comment]
Thanks @alexrp for the insight.
After thinking it over, I decided to using 'translate-c` to regenerated helper function definitions. What do you think?

**Comment by tw4452852** [comment]
Updated with @InKryption 's suggestion.

@alexrp There two things here to make BPF unusual:

- BPF helpers are just a dummy call, according to the [manual](https://man7.org/linux/man-pages/man7/bpf-helpers.7.html):

> Internally, eBPF programs call directly into the compiled helper
>        functions without requiring any foreign-function interface. As a
>        result, calling helpers introduces no overhead, thus offering
>        excellent performance.

- BPF programs are usually generated **w/o linking**, they are only being compiled(Currently no compiler supports linking BPF programs). Here's [the example from linux kernel]( https://github.com/torvalds/linux/blob/cfaaa7d010d1fc58f9717fcc8591201e741d2d49/samples/bpf/Makefile#L366).

**Review by alexrp — APPROVED** [review]

**Pull request merged**